libs/recache/src/handler.rs

```diff
@@ -0,0 +1,241 @@
+        )
+    }
+
+    pub fn default_request(&self) -> CacheRequestBuilder<C, K, V, SetRefreshAhead<SetTtl>>
+    where
+        Self: Sized,
+    {
+        self.request()
+            .ttl(self.default_ttl)
+            .maybe_refresh_ahead(self.default_refresh_ahead)
+    }
+}
+
+#[derive(Debug, Clone, thiserror::Error)]
+#[error("stringified error: {inner}")]
+pub struct CachedError {
+    inner: String,
+}
+
+impl CachedError {
+    pub fn new<E: std::fmt::Display>(e: E) -> Self {
+        Self {
+            inner: e.to_string(),
+        }
+    }
+}
+
+#[derive(Builder)]
+#[builder(finish_fn(vis = "", name = build_internal))]
+pub struct CacheRequest<C, K, V>
+where
+    C: AsyncCacheStore<K, V>,
+    K: Sync,
+{
+    #[builder(start_fn)]
+    cache: Arc<C>,
+    #[builder(start_fn)]
+    inflight: Arc<InflightMap<K, V, C>>,
+    #[builder(start_fn)]
+    on_hit: EventHandler<K, V>,
+    #[builder(start_fn)]
+    on_computed: EventHandler<K, V>,
+    #[builder(start_fn)]
+    on_inflight_computed: EventHandler<K, V>,
+    #[builder(start_fn)]
+    on_refresh_computed: EventHandler<K, V>,
+    #[builder(into)]
+    key: K,
+    ttl: Option<Duration>,
+    refresh_ahead: Option<Duration>,
+}
+
+#[derive(Error, Debug, Clone)]
+pub enum CacheRequestError<CacheError> {
+    #[error("compute error: {0}")]
+    ComputeError(String),
+    #[error("cache error: {0}")]
+    CacheError(#[from] CacheError),
+}
+
+impl<C, K, V> CacheRequest<C, K, V>
+where
+    C: AsyncCacheStore<K, V> + Send + Sync + 'static,
+    C::Error: Send + Sync + Clone,
+    K: Send + Sync + Clone + 'static + Hash + Eq,
+    V: Send + Sync + Clone + 'static,
+{
+    pub async fn execute<F, Fut, E>(self, compute: F) -> Result<V, CacheRequestError<C::Error>>
+    where
+        F: FnOnce() -> Fut,
+        Fut: Future<Output = Result<V, E>> + Send + 'static,
+        E: Display,
+    {
+        if let Some((val, maybe_ttl)) = self.cache.get_with_ttl(&self.key).await? {
+            (self.on_hit)(&self.key, &val);
+
+            if let Some(value_ttl) = maybe_ttl {
+                self.handle_refresh(compute, value_ttl);
+            }
+            return Ok(val);
+        }
+
+        let fut = match self.inflight.entry(self.key.clone()) {
+            dashmap::Entry::Occupied(entry) => {
+                let fut_shared = entry.get().clone();
+                async move {
+                    let val = fut_shared.await?;
+                    (self.on_inflight_computed)(&self.key, &val);
+                    Ok(val)
+                }
+                .boxed()
+            }
+            dashmap::Entry::Vacant(entry) => {
+                let fut_shared = Self::build_shared_future(compute());
+
+                entry.insert(fut_shared.clone());
+
+                let inflight = Arc::clone(&self.inflight);
+                async move {
+                    let res = fut_shared.await;
+                    if let Ok(val) = res.as_ref() {
```

Comment:
Nit: in commit message it was said that the `as_ref` was added to prevent potential memory leaks. Just interested, why there could be the leak if taken by value?

Reply:
`as_ref` was used just to emphasize that we don't need to consume the `val`. The fix itself was only about avoiding early return with `?`, so that we can always reach `inflight.remove(&self.key)` even if internal future returns an error. Also, I think there still could be a memory leak if `fut_shared` panics and we will not remove it from the map, but I guess we can just assume that users should provide panic-free functions.